### PR TITLE
Upstream a test for CustomElementRegistry interface from WebKit

### PR DESCRIPTION
**Pull request by rniwa**

### CODE REVIEW

**Comment by rniwa** [comment]
@kojiishi

File: custom-elements/CustomElementRegistry.html [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+        'customElements.define must throw a TypeError when the element interface is a string');
+    assert_throws({'name': 'TypeError'}, function () { customElements.define('badname', {}); },
+        'customElements.define must throw a TypeError when the element interface is an object');
+    assert_throws({'name': 'TypeError'}, function () { customElements.define('badname', []); },
+        'customElements.define must throw a TypeError when the element interface is an array');
+}, 'customElements.define must throw when the element interface is not a constructor');
+
+test(function () {
+    customElements.define('custom-html-element', HTMLElement);
+}, 'customElements.define must not throw the constructor is HTMLElement');
```
Comment (dominiccooney) [inline custom-elements/CustomElementRegistry.html]: Grammar: throw the constructor -> throw if the constructor

File: custom-elements/CustomElementRegistry.html [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+test(function () {
+    class SomeCustomElement extends HTMLElement {};
+
+    var calls = [];
+    var OtherCustomElement = new Proxy(class extends HTMLElement {}, {
+        get: function (target, name) {
+            calls.push(name);
+            return target[name];
+        }
+    })
```
Comment (dominiccooney) [inline custom-elements/CustomElementRegistry.html]: Maybe trailing semi here to not rely on automatic insertion?

File: custom-elements/CustomElementRegistry.html [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+
+    assert_false(unresolvedElement instanceof ConstructorCallsDefine);
+    assert_false(elementUpgradedDuringUpgrade instanceof DefinedDuringUpgrade);
+
+    customElements.define('constructor-calls-define', ConstructorCallsDefine);
+}, 'customElements.define unset the element definition is running flag before upgrading custom elements');
+
+(function () {
+    var testCase = async_test('customElements.define must not throw'
+        +' when defining another custom element in a different global object during Get(constructor, "prototype")', {timeout: 100});
```
Comment (dominiccooney) [inline custom-elements/CustomElementRegistry.html]: Style: maybe space between + and '

**Comment by domenic** [comment]
It's worth noting we have some tests already in https://github.com/w3c/web-platform-tests/blob/master/custom-elements/custom-element-registry/define.html. I'm fine with duplicate coverage (more tests better!), but we might want to consider the folder structure a bit more closely. E.g. maybe this should be split up into define-2.html, whenDefined.html, etc.? It's more important to land though so don't let that stop you.

**Comment by dominiccooney** [comment]
We have evolved defined.html downstream so it might be easier to upstream them as-is and do the merging upstream.

**Comment by sideshowbarker** [comment]
> We have evolved defined.html downstream so it might be easier to upstream them as-is and do the merging upstream.

OK, will merge this to master as-is then

**Pull request merged**